src/main/java/com/minecolonies/coremod/colony/buildings/BuildingWareHouse.java

```diff
@@ -0,0 +1,212 @@
+/**
+ * Class of the warehouse building.
+ */
+public class BuildingWareHouse extends AbstractBuilding
+{
+    /**
+     * String describing the Warehouse.
+     */
+    private static final String WAREHOUSE = "WareHouse";
+
+    /**
+     * Tag to store the deliverymen.
+     */
+    private static final String TAG_DELIVERYMAN = "Deliveryman";
+
+    /**
+     * The list of deliverymen registered to this building.
+     */
+    private static final List<Vec3d> registeredDeliverymen = new ArrayList<>();
+
+    /**
+     * The tileEntity of the building.
+     */
+    private TileEntityWareHouse tileEntity;
+
+    /**
+     * Instantiates a new warehouse building.
+     *
+     * @param c the colony.
+     * @param l the location
+     */
+    public BuildingWareHouse(final Colony c, final BlockPos l)
+    {
+        super(c, l);
+    }
+
+    /**
+     * Register deliveryman with the warehouse.
+     * @param buildingWorker the building of the worker.
+     * @return true if able to register or already registered
+     */
+    public boolean registerWithWareHouse(BuildingDeliveryman buildingWorker)
+    {
+        if(registeredDeliverymen.contains(new Vec3d(buildingWorker.getID())))
+        {
+            return true;
+        }
+
+        if(registeredDeliverymen.size() >= getBuildingLevel())
+        {
+            return false;
+        }
+
+        registeredDeliverymen.add(new Vec3d(buildingWorker.getID()));
+        return true;
+    }
+
+    /**
+     * Check if deliveryman is allowed to access warehouse.
+     * @param buildingWorker the building of the deliveryman.
+     * @return true if able to.
+     */
+    public boolean canAccessWareHouse(BuildingDeliveryman buildingWorker)
+    {
+        if(registeredDeliverymen.contains(new Vec3d(buildingWorker.getID())))
+        {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Get the deliverymen connected with this building.
+     * @return the unmodifiable List of positions of them.
+     */
+    public List<Vec3d> getRegisteredDeliverymen()
+    {
+        return Collections.unmodifiableList(registeredDeliverymen);
+    }
+
+    @NotNull
+    @Override
+    public String getSchematicName()
+    {
+        return WAREHOUSE;
+    }
+
+    @Override
+    public int getMaxBuildingLevel()
+    {
+        return 3;
+    }
+
+    @Override
+    public void readFromNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+
+        registeredDeliverymen.clear();
+        final NBTTagList deliverymanTagList = compound.getTagList(TAG_DELIVERYMAN, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < deliverymanTagList.tagCount(); i++)
+        {
+            final BlockPos pos = NBTUtil.getPosFromTag(deliverymanTagList.getCompoundTagAt(i));
+            if(getColony() != null && getColony().getBuilding(pos) instanceof AbstractBuildingWorker)
+            {
+                if(!registeredDeliverymen.contains(new Vec3d(pos)))
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1066)